### PR TITLE
formula_installer: skip bottle check on old Macs

### LINKED ISSUE #10832: Include dependencies when using install --build-from-source
### Provide a detailed description of the proposed feature

I'm using macOS 10.11 so I have to build all packages from source. I would expect that `brew install --build-from-source` would include dependencies. But it doesn't.

In the man page I see `Dependencies will still be installed from bottles if they are available.`, so this seems to be intended behaviour and why I created this as a feature request and not a bug.

Expected behaviour:

```
$ brew install --build-from-source ffmpeg
...
# ffmpeg installed
```

Actual behaviour:

```
$ brew install --build-from-source ffmpeg
...
Error: nasm: no bottle available!
You can try to install from source with:
  brew install --build-from-source nasm
...
$ brew install --build-from-source nasm
...
$ brew install --build-from-source ffmpeg
...
Error: pkg-config: no bottle available!
You can try to install from source with:
  brew install --build-from-source pkg-config
...
```

### What is the motivation for the feature?

1. Better user experience when installing packages from source
1. I would also imagine that this would match behaviour when not using `--build-from-source`, which I would expect does not require installation of all dependencies manually

### How will the feature be relevant to at least 90% of Homebrew users?

I don't know anything about the demographic breakdown of Homebrew users, but assuming 90% of Homebrew users are not running an unsupported version of macOS, this feature will likely not be relevant to them.

### What alternatives to the feature have been considered?

Another option might be to have some sort of additional parameter, e.g. `--include-dependencies` or `--build-dependencies-from-source` to allow enabling this behaviour explicitly.

Thanks!

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This makes sure the `no bottle available!` error only appears on macOS versions we support.

Alternatively, we could add something like a `--recursive-build-from-source` flag to `brew install`, which will build dependents from source too.

Fixes #10832.

### CODE REVIEW

**Comment by carlocab** [comment]
Ah, it looks like I can't just use an `OS::Mac` module like that. Whoops. It worked for me locally 😅

**Comment by Bo98** [comment]
The function isn't defined on Linux. Eiether you need to add a definition to `os/linux.rb` or avoid the call entirely outside macOS.

**Comment by Bo98** [comment]
Oh there's the generic OS case. Honestly, the compat methods living in `os/linux` has always felt weird and probably should have been in generic OS.

File: Library/Homebrew/formula_installer.rb [hunk 1]
```diff
@@ -231,6 +231,7 @@ def check_install_sanity
        # TODO: re-enable this on Linux when we merge linuxbrew-core into
        # homebrew-core and have full bottle coverage.
        (OS.mac? || ENV["CI"]) &&
+       !(OS.mac? && OS::Mac.outdated_release?) &&
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formula_installer.rb:234]: Would be nice to instead move this down to the 
```ruby
elsif !Homebrew::EnvConfig.developer? && (!installed_as_dependency? || !formula.any_version_installed?)
```

so instead it's something like
```ruby
elsif !Homebrew::EnvConfig.developer? && (!installed_as_dependency? || !formula.any_version_installed? || !OS.mac? || !OS::Mac.outdated_release?)
```

Comment (carlocab) [reply Library/Homebrew/formula_installer.rb]: Sounds good to me. Split it up a bit to readability. Though I will say I'm a bit more partial to `!(x && y)` instead of `(!x || !y)`, but happy to stay consistent with surrounding code.

File: Library/Homebrew/os/linux.rb [hunk 2]
```diff
@@ -47,6 +47,10 @@ def full_version
       Version::NULL
     end
 
+    def outdated_release?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/os/linux.rb:50]: Ideally let's fix the conditional so this isn't needed. Adjust the logic accordingly to fix.

Comment (carlocab) [reply Library/Homebrew/os/linux.rb:50]: My other idea was to define a constant `{HOMEBREW_}EARLIEST_SUPPORTED_MACOS` somewhere and use that, but I wasn't sure where to put it. Also in `formula_installer`?

Comment (Bo98) [reply Library/Homebrew/os/linux.rb:50]: Since we have `(!OS.mac? || !OS::Mac.outdated_release?)` (unless you are planning on changing that?) then this shouldn't be needed anymore.

Comment (carlocab) [reply Library/Homebrew/os/linux.rb:50]: I thought that too, but I think shellcheck complained about it. Maybe I was confused. Let me try again.

Comment (carlocab) [reply Library/Homebrew/os/linux.rb]: Ok, I was confused. Thanks for the nudge.

**Review by MikeMcQuaid — APPROVED** [review]
Thanks @carlocab!
